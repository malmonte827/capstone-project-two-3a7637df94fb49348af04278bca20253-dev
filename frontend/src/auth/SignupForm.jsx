import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import Alert from "../common/Alert"

/** Signup Form
 *
 * Shows form and handles update sto state on changes
 *
 * onSubmit calls signup function prop
 *
 * Routes -> SignuForm -> Alert
 * Routed as /signup
 */

function SignupForm({ signup }) {
    const [formData, setFormData] = useState({
        username: "",
        password: "",
        firstName: "",
        lastName: "",
        email: "",
        phoneNumber: "",
    });

    const [formErrors, setFormErrors] = useState([]);

    const navigate = useNavigate()


    /** Handles updating form fields on change */
    function handleChange(evt) {
        const { name, value } = evt.target;
        setFormData((data) => ({ ...data, [name]: value }));
    }

    /** handles form submition
     *
     * If succesful calls signup prop function and if not sets error
     */
    async function handleSubmit(evt) {
        evt.preventDefault();
        try {
            await signup(formData);
            navigate("/");
        } catch (err) {
            setFormErrors(err);
        }
    }

    return (
        <div className="SignupForm">
            <div className="container col-md-6 offset-md-3 col-lg-4 offset-lg-4">
                <h2 className="mb-3">Sign Up</h2>
                <div className="card">
                    <div className="card-body">
                        <form onSubmit={handleSubmit}>
                            <div className="mb-3">
                                <label
                                    className="form-label"
                                    htmlFor="username"
                                >
                                    Username
                                </label>
                                <input
                                    className="form-control"
                                    type="text"
                                    name="username"
                                    value={formData.username}
                                    onChange={handleChange}
                                    autoComplete="username"
                                />
                            </div>
                            <div className="mb-3">
                                <label
                                    className="form-label"
                                    htmlFor="password"
                                >
                                    Password
                                </label>
                                <input
                                    className="form-control"
                                    type="password"
                                    name="password"
                                    value={formData.password}
                                    onChange={handleChange}
                                    autoComplete="new-password"
                                />
                            </div>
                            <div className="mb-3">
                                <label
                                    className="form-label"
                                    htmlFor="firstName"
                                >
                                    First Name
                                </label>
                                <input
                                    className="form-control"
                                    type="text"
                                    name="firstName"
                                    value={formData.firstName}
                                    onChange={handleChange}
                                />
                            </div>
                            <div className="mb-3">
                                <label
                                    className="form-label"
                                    htmlFor="lastName"
                                >
                                    Last Name
                                </label>
                                <input
                                    className="form-control"
                                    type="text"
                                    name="lastName"
                                    value={formData.lastName}
                                    onChange={handleChange}
                                />
                            </div>
                            <div className="mb-3">
                                <label className="form-label" htmlFor="email">
                                    Email
                                </label>
                                <input
                                    className="form-control"
                                    type="email"
                                    name="email"
                                    value={formData.email}
                                    onChange={handleChange}
                                />
                            </div>
                            <div className="mb-3">
                                <label
                                    className="form-label"
                                    htmlFor="phoneNumber"
                                >
                                    Phone Number
                                </label>
                                <input
                                    className="form-control"
                                    type="tel"
                                    name="username"
                                    value={formData.phoneNumber}
                                    onChange={handleChange}
                                />
                            </div>

                            {formErrors.length ? (
                                <Alert type="danger" messages={formErrors} />
                            ) : null}
                            <div className="d-grid">
                                <button className="btn btn-primary">
                                    Submit
                                </button>
                            </div>
                        </form>
                    </div>
                </div>
            </div>
        </div>
    );
}

export default SignupForm;
