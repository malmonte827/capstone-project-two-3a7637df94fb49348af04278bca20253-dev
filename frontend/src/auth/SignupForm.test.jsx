import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom";
import { expect, vi } from "vitest";
import SignupForm from "./SignupForm";

vi.mock("../common/Alert");

describe("<SignupForm />", () => {
    const mockSignup = vi.fn();

    beforeEach(() => {
        mockSignup.mockReset();
    });

    it("renders without crashing", () => {
        render(
            <MemoryRouter>
                <SignupForm signup={mockSignup} />
            </MemoryRouter>
        );
    });

    it("matches snapshot", () => {
        const { asFragment } = render(
            <MemoryRouter>
                <SignupForm signup={mockSignup} />
            </MemoryRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("handles form input changes", () => {
        render(
            <MemoryRouter>
                <SignupForm signup={mockSignup} />
            </MemoryRouter>
        );

        const usernameInput = screen.getByLabelText("Username");
        const passwordInput = screen.getByLabelText("Password");
        const firstNameInput = screen.getByLabelText("First Name");
        const lastNameInput = screen.getByLabelText("Last Name");
        const emailInput = screen.getByLabelText("Email");
        const phoneInput = screen.getByLabelText("Phone Number");

        fireEvent.change(usernameInput, { target: { value: "testuser" } });
        fireEvent.change(passwordInput, { target: { value: "password123" } });
        fireEvent.change(firstNameInput, { target: { value: "John" } });
        fireEvent.change(lastNameInput, { target: { value: "Doe" } });
        fireEvent.change(emailInput, {
            target: { value: "johndoe@example.com" },
        });
        fireEvent.change(phoneInput, { target: { value: "123-456-7890" } });

        expect(usernameInput.value).toBe("testuser");
        expect(passwordInput.value).toBe("password123");
        expect(firstNameInput.value).toBe("John");
        expect(lastNameInput.value).toBe("Doe");
        expect(emailInput.value).toBe("johndoe@example.com");
        expect(phoneInput.value).toBe("123-456-7890");
    });

    it("submits the form and calls signup function", async () => {
        mockSignup.mockResolvedValueOnce();

        render(
            <MemoryRouter>
                <SignupForm signup={mockSignup} />
            </MemoryRouter>
        );

        fireEvent.change(screen.getByLabelText("Username"), {
            target: { value: "testuser" },
        });
        fireEvent.change(screen.getByLabelText("Password"), {
            target: { value: "password123" },
        });
        fireEvent.change(screen.getByLabelText("First Name"), {
            target: { value: "John" },
        });
        fireEvent.change(screen.getByLabelText("Last Name"), {
            target: { value: "Doe" },
        });
        fireEvent.change(screen.getByLabelText("Email"), {
            target: { value: "johndoe@example.com" },
        });
        fireEvent.change(screen.getByLabelText("Phone Number"), {
            target: { value: "123-456-7890" },
        });

        fireEvent.click(screen.getByText("Submit"));

        await waitFor(() => {
            expect(mockSignup).toHaveBeenCalledWith({
                username: "testuser",
                password: "password123",
                firstName: "John",
                lastName: "Doe",
                email: "johndoe@example.com",
                phoneNumber: "123-456-7890",
            });
        });
    });
});
