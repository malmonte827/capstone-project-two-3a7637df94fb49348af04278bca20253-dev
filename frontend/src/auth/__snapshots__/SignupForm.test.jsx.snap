// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SignupForm /> > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="SignupForm"
  >
    <div
      class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
    >
      <h2
        class="mb-3"
      >
        Sign Up
      </h2>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <form>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="username"
              >
                Username
              </label>
              <input
                autocomplete="username"
                class="form-control"
                id="username"
                name="username"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="password"
              >
                Password
              </label>
              <input
                autocomplete="new-password"
                class="form-control"
                id="password"
                name="password"
                type="password"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="firstName"
              >
                First Name
              </label>
              <input
                class="form-control"
                id="firstName"
                name="firstName"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="lastName"
              >
                Last Name
              </label>
              <input
                class="form-control"
                id="lastName"
                name="lastName"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="email"
              >
                Email
              </label>
              <input
                autocomplete="email"
                class="form-control"
                id="email"
                name="email"
                type="email"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="phoneNumber"
              >
                Phone Number
              </label>
              <input
                autocomplete="tel"
                class="form-control"
                id="phoneNumber"
                name="phoneNumber"
                type="tel"
                value=""
              />
            </div>
            <div
              class="d-grid"
            >
              <button
                class="btn btn-primary"
              >
                Submit
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
